File: vane-core/src/main/java/org/oddlama/vane/core/config/ConfigMaterialField.java
package org.oddlama.vane.core.config;

import static org.oddlama.vane.util.MaterialUtil.materialFrom;
import static org.oddlama.vane.util.StorageUtil.namespacedKey;

import java.lang.reflect.Field;
import java.util.function.Function;
import org.bukkit.Material;
import org.bukkit.configuration.file.YamlConfiguration;
import org.oddlama.vane.annotation.config.ConfigMaterial;
import org.oddlama.vane.core.YamlLoadException;

public class ConfigMaterialField extends ConfigField<Material> {

    public ConfigMaterial annotation;

    public ConfigMaterialField(
        Object owner,
        Field field,
        Function<String, String> mapName,
        ConfigMaterial annotation
    ) {
        super(owner, field, mapName, "material", annotation.desc());
        this.annotation = annotation;
    }

    @Override
    public Material def() {
        final var override = overriddenDef();
        if (override != null) {
            return override;
        } else {
            return annotation.def();
        }
    }

    @Override
    public boolean metrics() {
        final var override = overriddenMetrics();
        if (override != null) {
            return override;
        } else {
            return annotation.metrics();
        }
    }

    @Override
    public void generateYaml(StringBuilder builder, String indent, YamlConfiguration existingCompatibleConfig) {
        appendDescription(builder, indent);
        appendDefaultValue(
            builder,
            indent,
            "\"" + escapeYaml(def().getKey().getNamespace()) + ":" + escapeYaml(def().getKey().getKey()) + "\""
        );
        final var def = existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath())
            ? loadFromYaml(existingCompatibleConfig)
            : def();
        appendFieldDefinition(
            builder,
            indent,
            "\"" + escapeYaml(def.getKey().getNamespace()) + ":" + escapeYaml(def.getKey().getKey()) + "\""
        );
    }

    @Override
    public void checkLoadable(YamlConfiguration yaml) throws YamlLoadException {
        checkYamlPath(yaml);

        if (!yaml.isString(yamlPath())) {
            throw new YamlLoadException("Invalid type for yaml path '" + yamlPath() + "', expected string");
        }

        final var str = yaml.getString(yamlPath());
        final var split = str.split(":");
        if (split.length != 2) {
            throw new YamlLoadException(
                "Invalid material entry in list '" + yamlPath() + "': '" + str + "' is not a valid namespaced key"
            );
        }

        final var mat = materialFrom(namespacedKey(split[0], split[1]));
        if (mat == null) {
            throw new YamlLoadException(
                "Invalid material entry in list '" + yamlPath() + "': '" + str + "' does not exist"
            );
        }
    }

    public Material loadFromYaml(YamlConfiguration yaml) {
        final var split = yaml.getString(yamlPath()).split(":");
        return materialFrom(namespacedKey(split[0], split[1]));
    }

    public void load(YamlConfiguration yaml) {
        try {
            field.set(owner, loadFromYaml(yaml));
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Invalid field access on '" + field.getName() + "'. This is a bug.");
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/ConfigMaterialField.kt
package org.oddlama.vane.core.config

import org.bukkit.Material
import org.bukkit.configuration.file.YamlConfiguration
import org.oddlama.vane.annotation.config.ConfigMaterial
import org.oddlama.vane.core.YamlLoadException
import org.oddlama.vane.util.MaterialUtil.materialFrom
import org.oddlama.vane.util.StorageUtil.namespacedKey
import java.lang.reflect.Field

/**
 * Config field handler for material values.
 *
 * @param owner object containing the reflected config field.
 * @param field reflected config field.
 * @param mapName maps Java field names to YAML paths.
 * @param annotation source annotation metadata.
 */
class ConfigMaterialField(
    owner: Any?,
    field: Field,
    mapName: (String?) -> String?,
    /** Annotation metadata for this field. */
    var annotation: ConfigMaterial
) : ConfigField<Material?>(owner, field, mapName, "material", annotation.desc) {

    /** Returns the default value for this config field. */
    override fun def(): Material = overriddenDef() ?: annotation.def

    /** Returns whether metrics collection is enabled for this field. */
    override fun metrics(): Boolean = overriddenMetrics() ?: annotation.metrics

    /** Formats a material as a quoted namespaced key. */
    private fun Material.keyString() = "\"${escapeYaml(key.namespace)}:${escapeYaml(key.key)}\""

    /** Generates YAML for this field. */
    override fun generateYaml(builder: StringBuilder, indent: String, existingCompatibleConfig: YamlConfiguration?) {
        appendDescription(builder, indent)
        appendDefaultValue(builder, indent, def().keyString())
        val def = if (existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath()))
            loadFromYaml(existingCompatibleConfig) else def()
        appendFieldDefinition(builder, indent, def.keyString())
    }

    /** Validates that this field is loadable from YAML. */
    @Throws(YamlLoadException::class)
    override fun checkLoadable(yaml: YamlConfiguration) {
        checkYamlPath(yaml)
        if (!yaml.isString(yamlPath()))
            throw YamlLoadException("Invalid type for yaml path '${yamlPath()}', expected string")
        val str = yaml.getString(yamlPath())
        val (ns, key) = requireNamespacedKeyParts(yamlPath(), str, "material")
        materialFrom(namespacedKey(ns, key))
            ?: throw YamlLoadException("Invalid material entry in list '${yamlPath()}': '$str' does not exist")
    }

    /** Loads this field value from YAML. */
    fun loadFromYaml(yaml: YamlConfiguration): Material {
        val str = yaml.getString(yamlPath())
        val (ns, key) = requireNamespacedKeyParts(yamlPath(), str, "material")
        return materialFrom(namespacedKey(ns, key))!!
    }

    /** Writes the loaded value into the reflected field. */
    override fun load(yaml: YamlConfiguration) {
        try {
            field.set(owner, loadFromYaml(yaml))
        } catch (_: IllegalAccessException) {
            throw RuntimeException("Invalid field access on '${field.name}'. This is a bug.")
        }
    }
}
